### PR TITLE
React Redux Store setup

### PR DESCRIPTION
**Pull request by reshadbinharun**
image

* For components that need access to `store` state and/or write to `store`, we will need to add
- mapStateToProps
- mapDispatchToProps
... and wrap in connect(mapStateToProps, mapDispatchToProps)(ComponentClass) to allow it read/write access to store!

We do not necessarily need to use the store for all components.

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -20,6 +24,26 @@ export const PATHS = {
   register: "/register",
 }
 
+/*
+  STORE SETUP
+*/
+
+const mapStateToProps = state => {
```
Comment (reshadbinharun) [inline client/src/App.js:31]: The state should correspond to how `combineReducers` combines reducers, e.g. `countState` is the key that stores the store state change brought about by the `incrementReducer`

File: client/src/App.js [hunk 2]
```diff
@@ -20,6 +24,26 @@ export const PATHS = {
...
+  return ({
+    schoolName: state.schoolName,
+    count: state.countState.count
+  })
+}
+
+const mapDispatchToProps = dispatch => ({
```
Comment (reshadbinharun) [inline client/src/App.js:38]: A `dispatch` is used to trigger an action. The payload passed in is used to make changes to the store state in a reducer.

File: client/src/App.js [hunk 3]
```diff
@@ -20,6 +24,26 @@ export const PATHS = {
   register: "/register",
 }
 
+/*
```
Comment (reshadbinharun) [inline client/src/App.js:27]: Each component that needs to access/write to store, needs these mapState methods defined. These methods are passed in as arguments to `connect`

File: client/src/App.js [hunk 4]
```diff
@@ -190,14 +224,32 @@ export default class App extends Component {
...
           />
+          <Route exact path={`/playground`} render={(props) => 
+              <>
+                <Navbar
+                    navItems={alumniNavBarItems()}
+                    activeItem={'playground'}
+                />
+                <Button
+                  primary
+                  onClick={(e) => this.props.increment(1)}
```
Comment (reshadbinharun) [inline client/src/App.js:243]: dispatch methods defined in `mapDispatchToProps` are available as props.

File: client/src/App.js [hunk 5]
```diff
@@ -190,14 +224,32 @@ export default class App extends Component {
...
+                >
+                  Click to increase counter
+                </Button>
+                <Segment>
+                  Count in store is {this.props.count ? this.props.count : 0}
```
Comment (reshadbinharun) [inline client/src/App.js:248]: data defined in `mapStateToProps` is available as props.

File: client/src/App.js [hunk 6]
```diff
@@ -267,3 +319,5 @@ export default class App extends Component {
     )
   }
 }
+
+export default connect(mapStateToProps, mapDispatchToProps)(App);
```
Comment (reshadbinharun) [inline client/src/App.js:323]: The component must be connected in this manner, to allow access to state.

File: client/src/index.js [hunk 7]
```diff
@@ -1,8 +1,16 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
 import App from './App';
+import { Provider } from 'react-redux';
+import configureStore from "./redux/store"
 
 ReactDOM.render(
-        <App/>
+        <Provider
```
Comment (reshadbinharun) [inline client/src/index.js:8]: The Provider creates a store, and subscribes all nested children to the store.

File: client/src/redux/actions.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+export const testAction = () => dispatch => {
```
Comment (reshadbinharun) [inline client/src/redux/actions.js:1]: Actions are plain javascript objects. These methods are `action creators` in that they return actions. Actions are intended to represent a change that should happen to the store. They don't actually mutate the store in any way!

File: client/src/redux/actions.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export const testAction = () => {
```
Comment (reshadbinharun) [inline client/src/redux/actions.js:1]: Actions are plain javascript objects. These methods are action creators in that they return actions. Actions are intended to represent a change that should happen to the store. They don't actually mutate the store in any way!

File: client/src/redux/reducers.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { combineReducers } from 'redux';
```
Comment (reshadbinharun) [inline client/src/redux/reducers.js:1]: Reducers cause the actual mutations to the store. All reducers will `listen` for all actions, but only mutate for the action it is concerned with.

File: client/src/redux/reducers.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        case 'INCREMENT_ACTION':
+            return {
+                count: (state.count ? state.count : 0) + action.payload
+            }
+            default:
+                return state
+        }
+}
+
+export default combineReducers({
```
Comment (reshadbinharun) [inline client/src/redux/reducers.js:24]: The way combineReducers builds this object is import for `mapStateToProps`

**Comment by awilli70** [comment]
Looks awesome! Thanks for the documentation

**Pull request merged**